File: x/cpp/archive/cg.cpp
#include <string>

#include "util/MyTypes.h"
#include "util/MyBlas.h"
#include "util/MyNpy.h"
#include "util/MyTimer.h"

#include "cnpy.h" // npy

using namespace std;
using namespace MyTypes;
using namespace MyNpy;
using namespace MyBlas;
using namespace cnpy;

vector<double> cg(const Matrix &A, const vector<double> &b, const double epsilon);

int main() {
  string matrixFilePath = "../data/matrix.npy";
  string vectorFilePath = "../data/vector.npy";

  Matrix A = loadMatrix(matrixFilePath);
  vector<double> b = loadVector(vectorFilePath);

  const double epsilon = 1e-8;

  // timer
  std::chrono::system_clock::time_point  start, end;
  start = std::chrono::system_clock::now();
  auto x = cg(A, b, epsilon);
  end = std::chrono::system_clock::now();
  double elapsed = chrono::duration_cast<std::chrono::milliseconds>(end-start).count();
  cout << elapsed << "ms"<< endl;
  return 0;
}

vector<double> cg(const Matrix &A, const vector<double> &b, const double epsilon){
  const ui N = A.size();
  const ul max_iter = N * 2;
  const double b_norm = MyBlas::vecvec(b, b);
	vector<double> x(N, 0.0);

  // 初期残差
	vector<double> r = b - A * x;
	vector<double> p = r;
  double gamma = r*r;

  // 反復計算
  ul i = 0;

  while (i < max_iter)
  {
    // 収束判定
    double residual = vecvec(r, r) / b_norm;
    if (residual < epsilon) {
      break;
    }
    // 解の更新
    vector<double> v = matvec(A, p);
    double sigma = vecvec(p, v);
    double alpha = gamma / sigma;
    x = x + alpha * p;
    r = r - alpha * v;
    double old_gamma = gamma;
    gamma = vecvec(r, r);
    double beta = gamma / old_gamma;
    p = r + beta * p;
    i++;
  }

	return x;
}

File: v1/threads/cg.py
from .common import start, end as finish, init


def cg(A, b, epsilon, T, pu):
    if pu == 'cpu':
        from numpy import dot
        from numpy.linalg import norm
        x, b_norm, N, max_iter, residual, num_of_solution_updates = init(A, b, T, pu)
    else:
        from cupy import dot
        from cupy.linalg import norm
        A, b, x, b_norm, N, max_iter, residual, num_of_solution_updates = init(A, b, T, pu)

    # 初期残差
    r = b - dot(A, x)
    p = r.copy()
    gamma = dot(r, r)

    # 反復計算
    i = 0
    start_time = start(method_name=f'CG + {pu}')
    while i < max_iter:
        # 収束判定
        residual[i] = norm(r) / b_norm
        if residual[i] < epsilon:
            isConverged = True
            break

        # 解の更新
        v = dot(A, p)
        sigma = dot(p, v)
        alpha = gamma / sigma
        x += alpha * p
        r -= alpha * v
        old_gamma = gamma.copy()
        gamma = dot(r, r)
        beta = gamma / old_gamma
        p = r + beta * p
        i += 1
        num_of_solution_updates[i] = i
    else:
        isConverged = False

    elapsed_time = finish(start_time, isConverged, i, residual[i])
    return elapsed_time, num_of_solution_updates[:i+1], residual[:i+1]
